pkg/controller/constraint/constraint_controller.go

```diff
@@ -333,6 +282,15 @@ func (r *ReconcileConstraint) Reconcile(ctx context.Context, request reconcile.R
 		}
 	}()
 
+	ct := &v1beta1.ConstraintTemplate{}
+	err = r.reader.Get(ctx, types.NamespacedName{Name: strings.ToLower(instance.GetKind())}, ct)
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+
+	generateVAPB := *DefaultGenerateVAPB && HasVAPCel(ct)
```

Comment:
I thought we were going to get rid of the condition of only generating the binding if the constraint template had VAP CEL?

Link to relevant chat thread: https://openpolicyagent.slack.com/archives/GQK16NZEJ/p1719537610076419?thread_ts=1719532061.611639&cid=GQK16NZEJ

The thread mentions potentially adding a warning if we are creating a binding for a template that does not have a VAP implementation.

IIRC we are able to do this b/c users should be able to use `scopedEnforcementActions` if they need to modify behavior more granularly.

Reply:
I am skeptical. I suspect we are trying to be too clever here and that we will shoot ourselves in the foot. Particularly if users want to leverage fail-closed. I suppose the behavior could be customized further down the line with flags.

What happens when users specify "only use VAP" using scopedEA for a rego-only template? We continue to fail to create the binding and add a warning?

Looks like we are already triggering re-reconciles of all constraints when templates get updated, which is good:

https://github.com/open-policy-agent/gatekeeper/blob/0f2048422a81519aa7cfc8363f408a68c02938d8/pkg/controller/constrainttemplate/constrainttemplate_controller.go#L699-L717

If none of these concerns are blocking, SGTM

Reply:
> Particularly if users want to leverage fail-closed.

@maxsmythe what do you mean by this? can you elaborate?

> What happens when users specify "only use VAP" using scopedEA for a rego-only template? We continue to fail to create the binding and add a warning?

I think yes 🤔 . There wouldn't be any enforcement anyways in this situation given VAP won't be there even if we end up creating VAPB otherwise. We should follow up with VAP warning and error message schema for constraint status after this PR.

Reply:
> Particularly if users want to leverage fail-closed.

Are you referring to when user wants to use vap and vapb to enforce but the CT only has rego, GK should fail instead of warn and let GK webhook handle it? If so, then I agree if user specified `scopedEA` to use VAP but CT doesnt have CEL, then we should return error when constraint is created or updated. However, if there is no constraint level `scopedEA` provided, we should not break all constraints that do not have CEL in the CT.